### ISSUE TITLE
Implement sd_notify on linux

### ISSUE THREAD

**Issue by vodif** [storj/storj#5959]
**Describe the feature or problem you'd like to solve**
I'm running the storagenode binary with systemd with the following .service file:

```
[Unit]
Description  = Storage Node service
After        = default.target

[Service]
Type         = simple
User         = storj
Group        = storj
ExecStart    = /home/storj/bin/storagenode run --identity-dir "/home/storj/identity/storagenode" --config-dir "/home/storj/config"
Restart      = always
RestartSec   = 5
NotifyAccess = main
CPUWeight    = 50
CPUQuota     = 50%
MemoryHigh   = 400M
MemoryMax    = 600M
MemorySwapMax= 0

[Install]
WantedBy     = default.target
```

and this is the second time I wake up to emails stating "Your Node has Gone Offline".
Upon investigating I realize that the same issue occured as last time. The `storagenode` binary is unresponsive. Not even handling os signals. When I try to stop the storagenode.service I get:
```
Jun 11 09:16:23 localhost systemd[1]: storagenode.service: State 'stop-sigterm' timed out. Killing.
Jun 11 09:16:23 localhost systemd[1]: storagenode.service: Killing process 3578496 (storagenode) with signal SIGKILL.
Jun 11 09:16:23 localhost systemd[1]: storagenode.service: Killing process 3578497 (storagenode) with signal SIGKILL.
...
Jun 11 09:16:23 localhost systemd[1]: storagenode.service: Killing process 3989623 (n/a) with signal SIGKILL.
Jun 11 09:16:23 localhost systemd[1]: storagenode.service: Main process exited, code=killed, status=9/KILL
Jun 11 09:16:23 localhost systemd[1]: storagenode.service: Failed with result 'timeout'.
Jun 11 09:16:23 localhost systemd[1]: storagenode.service: Consumed 59min 25.053s CPU time.
```

Which means your binary was properly stuck.

This is a common issue with software written in Go. See this one for example: https://github.com/authelia/authelia/issues/4963

**Propose a Solution**
(There probably exists a similar solution on windows but honestly I don't give a damn about microsoft.)
Implement https://www.freedesktop.org/software/systemd/man/sd_notify.html, so in cases like the above one, at least systemd would kick your software in the butt.

This should not take more than a couple of lines of code (one READY=1 at start and one WATCHDOG=1 in your main loop) and although I don't know how you run your binary inside the docker container, but for those of us who don't/won't/don't want to use docker, it would make it so that any other bugs you have with your main loop, at least it would be restarted by systemd.

**Additional context**
And while I'm at it... please for the love of ... the actual 2 people of the 20k node operators who actually care about their system and they monitor their logs, please STOP changing your logformat in every flippin' release.
It is utmost annoying when at a random time storagenode-updater updates storagenode and then a TON utterly useless logs are arriving from the logcheck system just because  - for example - in your last release of v1.79.4 in the log:
```
Jun  7 17:25:12 localhost storagenode[2086006]: 2023-06-07T17:25:12.009+0200#011ERROR#011piecestore#011upload failed#011{"process": "storagenode", "Piece ID": "QZNWVJYSM62HSZTWQIEMCHBWHXRQUJITCF2GABB23AFU7HRVD26A", "Satellite ID": "12EayRS2V1kEsWESU9QMRseFhdxYxKicsiFmxrsLZHeLUtdps3S", "Action": "PUT", "error": "context canceled", "errorVerbose": "context canceled\n\tstorj.io/common/rpc/rpcstatus.Wrap:75\n\tstorj.io/storj/storagenode/piecestore.(*Endpoint).Upload.func5:498\n\tstorj.io/storj/storagenode/piecestore.(*Endpoint).Upload:504\n\tstorj.io/common/pb.DRPCPiecestoreDescription.Method.func1:243\n\tstorj.io/drpc/drpcmux.(*Mux).HandleRPC:33\n\tstorj.io/common/rpc/rpctracing.(*Handler).HandleRPC:61\n\tstorj.io/common/experiment.(*Handler).HandleRPC:42\n\tstorj.io/drpc/drpcserver.(*Server).handleRPC:124\n\tstorj.io/drpc/drpcserver.(*Server).ServeOne:66\n\tstorj.io/drpc/drpcserver.(*Server).Serve.func2:114\n\tstorj.io/drpc/drpcctx.(*Tracker).track:35", "Size": 52224, "Remote Address": "216.66.40.82:55494"}
```

you changed the word `Process` to `process`.

**Comment by thepaul** [comment]
As far as I know, though, anytime the storagenode binary stops responding for too long to signals and/or network queries, it's because it has been asked to perform more disk IO than the system can handle. (This is the case when the storagenode binary appears to be stuck in status `D` in `ps aux`.) And restarting the binary won't help with that. It will lessen the amount of I/O temporarily while it gets back up and running and handling new requests, but it is likely to reach the same level of load again very soon, particularly because on start it launches a task to traverse the storage area, counting how many bytes are stored per satellite.

So, in short, we don't recommend automatically restarting when the storagenode service is unresponsive for some time.

**Comment by vodif** [comment]
Fortunately I'm here to defy your beliefs.

1. It is not accidental that I've linked a completely different - although written in the same language - software bug, since that software has _exactly_ 0 disk io, but still has (had?) the exact same bug yours have.
The issue I'm talking about here is when `storagenode` has exactly 0 disk io (I'm monitoring it, I know) but still not responding to signals. (Yes, there can be other issues/hardware errors when a userspace binary stops responding but that also involves kernel error logs).

2. Restarting the `storagenode` does help. I know, because I've done it.. now twice unfortunately, as I also said this before. (Indeed there is that pesky filewalker, which can be disabled to run at startup, but your other trash chore is still running at startup and at every 24h without the ability to turn it off, but this is completely unrelated to this issue)

> So, in short, we don't recommend automatically restarting when the storagenode service is unresponsive for some time.

I'm a software engineer. If I ever recommended something like this to our customers I wouldn't be working at this company for long that's for sure.

**Comment by thepaul** [comment]
Ok, this is good to know. To my knowledge, you are the first person to report this behavior (process wholly unresponsive, no disk i/o). There may be something new we can learn from this.

Could you collect more information the next time this happens: what does the CPU load look like, what does `ptrace -f -o node-ptrace.log -p $nodepid` report in connection with `lsof -n -p $nodepid`, etc. A core dump could help greatly. Or, stack dumps would be easier to transmit. If you kill the process with SIGABRT, it may be alive enough to emit all goroutine stacks to stderr.
